File: apps/pop_it/pop_it.py
"""
apps/pop_it/pop_it.py — "Pop It" mini-game.

Players use their index finger to pop falling colored balls within a 30-second
session, then view their score and choose to replay or exit.
"""

import cv2
import time
import random
import math

from ..base import BaseApp

# ── Constants ─────────────────────────────────────────────────────────────────

_GAME_DURATION = 30          # seconds

_BALL_COLORS = [             # BGR
    (0,   80, 255),          # red
    (0,  165, 255),          # orange
    (0,  230, 255),          # yellow
    (30, 255, 120),          # green
    (255, 200,  20),         # cyan
    (255,  60,  60),         # blue
    (200,   0, 200),         # purple
    (255,   0, 160),         # pink
    (160, 255,   0),         # lime
]

_SPEED_MIN    = 120          # px / s
_SPEED_MAX    = 340
_RADIUS_MIN   = 28
_RADIUS_MAX   = 58
_SPAWN_MIN    = 0.5          # seconds between ball spawns
_SPAWN_MAX    = 1.5
_POP_DURATION = 0.35         # seconds a pop-burst lasts
_BTN_COOLDOWN = 0.75         # seconds between button triggers


# ── Internal data classes ─────────────────────────────────────────────────────

class _Ball:
    __slots__ = ("x", "y", "radius", "speed", "color")

    def __init__(self, x, y, radius, speed, color):
        self.x      = float(x)
        self.y      = float(y)
        self.radius = radius
        self.speed  = speed
        self.color  = color


class _PopEffect:
    __slots__ = ("x", "y", "radius", "color", "born")

    def __init__(self, x, y, radius, color):
        self.x      = x
        self.y      = y
        self.radius = radius
        self.color  = color
        self.born   = time.time()


# ── App ───────────────────────────────────────────────────────────────────────

class PopItGame(BaseApp):
    """Pop It mini-game rendered as an OpenCV overlay."""

    def __init__(self):
        self._state      = "idle"   # "idle" | "active" | "end"
        self.score       = 0
        self._balls:   list[_Ball]      = []
        self._effects: list[_PopEffect] = []
        self._game_start  = 0.0
        self._next_spawn  = 0.0
        self._last_tick   = 0.0
        self._btn_times: dict[str, float] = {}

    # ── BaseApp interface ─────────────────────────────────────────────────────

    @property
    def is_idle(self) -> bool:
        return self._state == "idle"

    def process_fingertip(self, fx: int, fy: int, w: int, h: int) -> None:
        self._check_buttons(fx, fy, w, h)
        if self._state == "active":
            self._try_pop(fx, fy)

    def update_and_draw(self, frame) -> None:
        h, w   = frame.shape[:2]
        now    = time.time()
        dt     = now - self._last_tick if self._last_tick else 0.0
        self._last_tick = now

        if self._state == "active":
            self._update(now, dt, w, h)
            self._draw_active(frame, w, h, now)
        elif self._state == "end":
            self._draw_end(frame, w, h)

        # Trigger button is always drawn in idle
        if self._state == "idle":
            self._draw_button(frame, w, h)

    # ── Game flow ─────────────────────────────────────────────────────────────

    def _start(self):
        now = time.time()
        self._state      = "active"
        self.score       = 0
        self._balls      = []
        self._effects    = []
        self._game_start = now
        self._next_spawn = now + random.uniform(_SPAWN_MIN, _SPAWN_MAX)
        self._last_tick  = now

    def _end(self):
        self._state  = "end"
        self._balls  = []

    def _exit(self):
        self._state   = "idle"
        self._balls   = []
        self._effects = []

    # ── Update ────────────────────────────────────────────────────────────────

    def _update(self, now: float, dt: float, w: int, h: int):
        if now - self._game_start >= _GAME_DURATION:
            self._end()
            return

        # Spawn a new ball
        if now >= self._next_spawn:
            r = random.randint(_RADIUS_MIN, _RADIUS_MAX)
            self._balls.append(_Ball(
                x      = random.randint(r, w - r),
                y      = -r,
                radius = r,
                speed  = random.uniform(_SPEED_MIN, _SPEED_MAX),
                color  = random.choice(_BALL_COLORS),
            ))
            self._next_spawn = now + random.uniform(_SPAWN_MIN, _SPAWN_MAX)

        # Move balls, remove those that have left the screen
        alive = []
        for b in self._balls:
            b.y += b.speed * dt
            if b.y - b.radius < h:
                alive.append(b)
        self._balls = alive

        # Expire old pop effects
        self._effects = [e for e in self._effects
                         if now - e.born < _POP_DURATION]

    # ── Interaction ───────────────────────────────────────────────────────────

    def _try_pop(self, fx: int, fy: int):
        for ball in self._balls:
            if math.hypot(fx - ball.x, fy - ball.y) <= ball.radius:
                self._balls.remove(ball)
                self._effects.append(
                    _PopEffect(ball.x, ball.y, ball.radius, ball.color)
                )
                self.score += 1
                return  # one pop per frame per finger

    def _check_buttons(self, fx: int, fy: int, w: int, h: int):
        now = time.time()

        def _ok(name: str) -> bool:
            if now - self._btn_times.get(name, 0) >= _BTN_COOLDOWN:
                self._btn_times[name] = now
                return True
            return False

        if self._state == "idle":
            if self._in(fx, fy, *self._btn_popit(w, h)) and _ok("start"):
                self._start()
        elif self._state == "end":
            if self._in(fx, fy, *self._btn_restart(w, h)) and _ok("restart"):
                self._start()
            elif self._in(fx, fy, *self._btn_exit(w, h)) and _ok("exit"):
                self._exit()

    @staticmethod
    def _in(fx, fy, x1, y1, x2, y2) -> bool:
        return x1 <= fx <= x2 and y1 <= fy <= y2

    # ── Button rects (x1, y1, x2, y2) ────────────────────────────────────────

    @staticmethod
    def _btn_popit(w, h):
        return (w - 160, 12, w - 12, 66)

    @staticmethod
    def _btn_restart(w, h):
        cx, cy = w // 2, h // 2 + 60
        return (cx - 150, cy - 28, cx - 10, cy + 28)

    @staticmethod
    def _btn_exit(w, h):
        cx, cy = w // 2, h // 2 + 60
        return (cx + 10, cy - 28, cx + 150, cy + 28)

    # ── Drawing ───────────────────────────────────────────────────────────────

    def _draw_button(self, frame, w, h):
        x1, y1, x2, y2 = self._btn_popit(w, h)
        cv2.rectangle(frame, (x1, y1), (x2, y2), (30, 160, 255), -1)
        cv2.rectangle(frame, (x1, y1), (x2, y2), (255, 255, 255), 2)
        cv2.putText(frame, "Pop It!", (x1 + 10, y1 + 38),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.78, (255, 255, 255), 2)

    def _draw_active(self, frame, w, h, now: float):
        # Balls
        for ball in self._balls:
            x, y, r = int(ball.x), int(ball.y), ball.radius
            cv2.circle(frame, (x, y), r, ball.color, -1)
            cv2.circle(frame, (x, y), r, (255, 255, 255), 2)

        # Pop burst rings
        for e in self._effects:
            progress = (now - e.born) / _POP_DURATION
            for ring in range(3):
                r     = int(e.radius + ring * 14 + progress * 55)
                alpha = max(0.0, 1.0 - progress - ring * 0.25)
                if alpha <= 0:
                    continue
                overlay = frame.copy()
                cv2.circle(overlay, (int(e.x), int(e.y)), r, e.color, 3)
                cv2.addWeighted(overlay, alpha, frame, 1 - alpha, 0, frame)

        # Timer — top-center
        remaining   = max(0.0, _GAME_DURATION - (now - self._game_start))
        secs        = int(math.ceil(remaining))
        timer_color = (0, 255, 255) if remaining > 10 else (0, 90, 255)
        timer_txt   = f"{secs}s"
        (tw, _), _  = cv2.getTextSize(timer_txt, cv2.FONT_HERSHEY_SIMPLEX, 1.5, 3)
        cv2.putText(frame, timer_txt, (w // 2 - tw // 2, 55),
                    cv2.FONT_HERSHEY_SIMPLEX, 1.5, timer_color, 3)

        # Score — top-left
        score_txt = f"Score: {self.score}"
        cv2.rectangle(frame, (10, 10), (220, 62), (0, 0, 0), -1)
        cv2.rectangle(frame, (10, 10), (220, 62), (50, 220, 100), 2)
        cv2.putText(frame, score_txt, (20, 48),
                    cv2.FONT_HERSHEY_SIMPLEX, 1.1, (50, 255, 120), 2)

    def _draw_end(self, frame, w, h):
        # Dark overlay
        overlay = frame.copy()
        cv2.rectangle(overlay, (0, 0), (w, h), (0, 0, 0), -1)
        cv2.addWeighted(overlay, 0.55, frame, 0.45, 0, frame)

        cx, cy = w // 2, h // 2

        # Title
        title      = "Game Over!"
        (tw, _), _ = cv2.getTextSize(title, cv2.FONT_HERSHEY_SIMPLEX, 2.2, 4)
        cv2.putText(frame, title, (cx - tw // 2, cy - 80),
                    cv2.FONT_HERSHEY_SIMPLEX, 2.2, (0, 240, 255), 4)

        # Score
        score_txt  = f"Score: {self.score}"
        (sw, _), _ = cv2.getTextSize(score_txt, cv2.FONT_HERSHEY_SIMPLEX, 1.6, 3)
        cv2.putText(frame, score_txt, (cx - sw // 2, cy - 10),
                    cv2.FONT_HERSHEY_SIMPLEX, 1.6, (255, 255, 255), 3)

        # Restart button
        x1, y1, x2, y2 = self._btn_restart(w, h)
        cv2.rectangle(frame, (x1, y1), (x2, y2), (30, 180, 30), -1)
        cv2.rectangle(frame, (x1, y1), (x2, y2), (255, 255, 255), 2)
        lbl        = "Restart"
        (lw, _), _ = cv2.getTextSize(lbl, cv2.FONT_HERSHEY_SIMPLEX, 0.85, 2)
        cv2.putText(frame, lbl, (x1 + (x2 - x1 - lw) // 2, y1 + 38),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.85, (255, 255, 255), 2)

        # Exit button
        x1, y1, x2, y2 = self._btn_exit(w, h)
        cv2.rectangle(frame, (x1, y1), (x2, y2), (30, 30, 200), -1)
        cv2.rectangle(frame, (x1, y1), (x2, y2), (255, 255, 255), 2)
        lbl        = "Exit"
        (lw, _), _ = cv2.getTextSize(lbl, cv2.FONT_HERSHEY_SIMPLEX, 0.85, 2)
        cv2.putText(frame, lbl, (x1 + (x2 - x1 - lw) // 2, y1 + 38),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.85, (255, 255, 255), 2)

        # Prompt
        prompt     = "Touch a button with your index finger"
        (pw, _), _ = cv2.getTextSize(prompt, cv2.FONT_HERSHEY_SIMPLEX, 0.6, 1)
        cv2.putText(frame, prompt, (cx - pw // 2, cy + 120),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.6, (180, 180, 180), 1)

File: docs/js/apps/pop_it.js
/**
 * apps/pop_it.js — "Pop It" mini-game.
 *
 * Direct port of apps/pop_it/pop_it.py.
 * cv2 drawing → Canvas 2D API.
 * time.time()  → performance.now() / 1000  (seconds).
 * BGR colors   → RGB  (channels reversed).
 */

// ── Constants ────────────────────────────────────────────────────────────────

const GAME_DURATION = 30;     // seconds

const BALL_COLORS = [
  '#FF5020', '#FFA800', '#FFE500',
  '#78FF1E', '#14C8FF', '#3C3CFF',
  '#C800C8', '#A000FF', '#00FFA0',
];

const SPEED_MIN    = 120;   // px/s
const SPEED_MAX    = 340;
const RADIUS_MIN   = 28;
const RADIUS_MAX   = 58;
const SPAWN_MIN    = 0.5;   // seconds between spawns
const SPAWN_MAX    = 1.5;
const POP_DURATION = 0.35;  // seconds a burst lasts
const BTN_COOLDOWN = 0.75;  // seconds between button triggers

// ── Helpers ──────────────────────────────────────────────────────────────────

function rand(min, max) { return min + Math.random() * (max - min); }
function randInt(min, max) { return Math.floor(rand(min, max + 1)); }
function pick(arr) { return arr[Math.floor(Math.random() * arr.length)]; }
function now() { return performance.now() / 1000; }

// ── UI helpers ───────────────────────────────────────────────────────────────

/** Filled rounded-rect button with centered label. */
function drawBtn(ctx, x1, y1, x2, y2, label, bg, options = {}) {
  const { r = 14, border = 'rgba(255,255,255,0.25)', fontSize = 20, bold = true } = options;
  const bw = x2 - x1, bh = y2 - y1;
  ctx.save();
  ctx.shadowColor   = 'rgba(0,0,0,0.45)';
  ctx.shadowBlur    = 10;
  ctx.shadowOffsetY = 3;
  ctx.beginPath();
  ctx.roundRect(x1, y1, bw, bh, r);
  ctx.fillStyle = bg;
  ctx.fill();
  ctx.shadowColor = 'transparent';
  ctx.strokeStyle = border;
  ctx.lineWidth   = 1.5;
  ctx.stroke();
  ctx.font          = `${bold ? 'bold ' : ''}${fontSize}px system-ui, sans-serif`;
  ctx.fillStyle     = '#fff';
  ctx.textAlign     = 'center';
  ctx.textBaseline  = 'middle';
  ctx.fillText(label, x1 + bw / 2, y1 + bh / 2);
  ctx.restore();
}

/** Dark glass panel. */
function drawPanel(ctx, x, y, w, h, r = 16) {
  ctx.save();
  ctx.shadowColor   = 'rgba(0,0,0,0.5)';
  ctx.shadowBlur    = 12;
  ctx.shadowOffsetY = 3;
  ctx.beginPath();
  ctx.roundRect(x, y, w, h, r);
  ctx.fillStyle = 'rgba(12, 12, 22, 0.72)';
  ctx.fill();
  ctx.shadowColor = 'transparent';
  ctx.strokeStyle = 'rgba(255,255,255,0.12)';
  ctx.lineWidth   = 1;
  ctx.stroke();
  ctx.restore();
}

// ── PopItGame ─────────────────────────────────────────────────────────────────

export class PopItGame {
  constructor() {
    this._state     = 'idle';   // 'idle' | 'active' | 'end'
    this.score      = 0;
    this._balls     = [];
    this._effects   = [];
    this._gameStart = 0;
    this._nextSpawn = 0;
    this._lastTick  = 0;
    this._btnTimes  = {};
  }

  // ── BaseApp interface ───────────────────────────────────────────────────────

  get isIdle() { return this._state === 'idle'; }

  /** Called by the sidebar dock to launch this app. */
  open() {
    if (this._state !== 'idle') return;
    this._start();
  }

  processFingertip(fx, fy, w, h) {
    this._checkButtons(fx, fy, w, h);
    if (this._state === 'active') this._tryPop(fx, fy);
  }

  processHand(_landmarks, _w, _h) { /* not used by this app */ }

  updateAndDraw(ctx, w, h, anyActive = false) {
    const t  = now();
    const dt = this._lastTick ? t - this._lastTick : 0;
    this._lastTick = t;

    if (this._state === 'active') {
      this._update(t, dt, w, h);
      this._drawActive(ctx, w, h, t);
    } else if (this._state === 'end') {
      this._drawEnd(ctx, w, h);
    }

    // Hide the trigger button while another app is running
    if (this._state === 'idle' && !anyActive) {
      this._drawIdleButton(ctx, w, h);
    }
  }

  // ── Game flow ───────────────────────────────────────────────────────────────

  _start() {
    const t        = now();
    this._state    = 'active';
    this.score     = 0;
    this._balls    = [];
    this._effects  = [];
    this._gameStart = t;
    this._nextSpawn = t + rand(SPAWN_MIN, SPAWN_MAX);
    this._lastTick  = t;
  }

  _end()  { this._state = 'end';  this._balls = []; }

  _exit() {
    this._state   = 'idle';
    this._balls   = [];
    this._effects = [];
    // Prevent the idle trigger from firing immediately while the finger is
    // still hovering over the exit button area.
    this._btnTimes['start'] = now();
  }

  // ── Update ──────────────────────────────────────────────────────────────────

  _update(t, dt, w, h) {
    if (t - this._gameStart >= GAME_DURATION) { this._end(); return; }

    if (t >= this._nextSpawn) {
      const r = randInt(RADIUS_MIN, RADIUS_MAX);
      this._balls.push({
        x: rand(r, w - r),
        y: -r,
        radius: r,
        speed: rand(SPEED_MIN, SPEED_MAX),
        color: pick(BALL_COLORS),
      });
      this._nextSpawn = t + rand(SPAWN_MIN, SPAWN_MAX);
    }

    this._balls = this._balls.filter(b => {
      b.y += b.speed * dt;
      return b.y - b.radius < h;
    });

    this._effects = this._effects.filter(e => t - e.born < POP_DURATION);
  }

  // ── Interaction ─────────────────────────────────────────────────────────────

  _tryPop(fx, fy) {
    for (let i = 0; i < this._balls.length; i++) {
      const b = this._balls[i];
      if (Math.hypot(fx - b.x, fy - b.y) <= b.radius) {
        this._effects.push({ x: b.x, y: b.y, radius: b.radius, color: b.color, born: now() });
        this._balls.splice(i, 1);
        this.score++;
        return;
      }
    }
  }

  _checkButtons(fx, fy, w, h) {
    const t  = now();
    const ok = (name) => {
      if (t - (this._btnTimes[name] || 0) >= BTN_COOLDOWN) {
        this._btnTimes[name] = t;
        return true;
      }
      return false;
    };

    if (this._state === 'active') {
      if (this._inRect(fx, fy, ...this._btnActiveExit(w, h)) && ok('exit')) this._exit();
    } else if (this._state === 'end') {
      if (this._inRect(fx, fy, ...this._btnRestart(w, h)) && ok('restart')) this._start();
      else if (this._inRect(fx, fy, ...this._btnExit(w, h)) && ok('exit')) this._exit();
    }
  }

  _inRect(fx, fy, x1, y1, x2, y2) {
    return fx >= x1 && fx <= x2 && fy >= y1 && fy <= y2;
  }

  // ── Button rects [x1, y1, x2, y2] ──────────────────────────────────────────

  _btnPopit(w, _h)       { return [w - 160, 12, w - 12, 66]; }
  _btnActiveExit(_w, _h) { return [12, 12, 120, 56]; }  // top-LEFT — away from idle trigger

  _btnRestart(w, h) {
    const cx = w / 2, cy = h / 2 + 60;
    return [cx - 150, cy - 28, cx - 10, cy + 28];
  }

  _btnExit(w, h) {
    const cx = w / 2, cy = h / 2 + 60;
    return [cx + 10, cy - 28, cx + 150, cy + 28];
  }

  // ── Drawing ─────────────────────────────────────────────────────────────────

  _drawIdleButton(ctx, w, h) {
    const [x1, y1, x2, y2] = this._btnPopit(w, h);
    drawBtn(ctx, x1, y1, x2, y2, 'Pop It!', 'rgba(230, 110, 20, 0.88)');
  }

  _drawActive(ctx, w, h, t) {
    // ── Balls ──────────────────────────────────────────────────────────────
    for (const b of this._balls) {
      ctx.save();
      // Radial gradient for depth
      const grad = ctx.createRadialGradient(
        b.x - b.radius * 0.3, b.y - b.radius * 0.35, b.radius * 0.1,
        b.x, b.y, b.radius,
      );
      grad.addColorStop(0, '#fff');
      grad.addColorStop(0.3, b.color);
      grad.addColorStop(1, b.color + '88');
      ctx.beginPath();
      ctx.arc(b.x, b.y, b.radius, 0, Math.PI * 2);
      ctx.fillStyle = grad;
      ctx.fill();
      ctx.strokeStyle = 'rgba(255,255,255,0.35)';
      ctx.lineWidth   = 1.5;
      ctx.stroke();
      ctx.restore();
    }

    // ── Pop burst rings ────────────────────────────────────────────────────
    for (const e of this._effects) {
      const progress = (t - e.born) / POP_DURATION;
      for (let ring = 0; ring < 3; ring++) {
        const alpha = Math.max(0, 1 - progress - ring * 0.25);
        if (alpha <= 0) continue;
        const r = e.radius + ring * 14 + progress * 55;
        ctx.save();
        ctx.globalAlpha = alpha;
        ctx.beginPath();
        ctx.arc(e.x, e.y, r, 0, Math.PI * 2);
        ctx.strokeStyle = e.color;
        ctx.lineWidth   = 3;
        ctx.stroke();
        ctx.restore();
      }
    }

    // ── Score + Timer — combined panel, top-center ─────────────────────────
    const remaining = Math.max(0, GAME_DURATION - (t - this._gameStart));
    const urgent    = remaining <= 10;
    const timerText = `${Math.ceil(remaining)}s`;
    const panelW = 320, panelH = 58;
    const panelX = w / 2 - panelW / 2;
    drawPanel(ctx, panelX, 8, panelW, panelH, 14);

    // Score (left half)
    ctx.save();
    ctx.font         = 'bold 26px system-ui, sans-serif';
    ctx.fillStyle    = '#30E88A';
    ctx.textAlign    = 'left';
    ctx.textBaseline = 'middle';
    ctx.fillText(`Score: ${this.score}`, panelX + 16, 37);
    ctx.restore();

    // Divider
    ctx.save();
    ctx.strokeStyle = 'rgba(255,255,255,0.15)';
    ctx.lineWidth   = 1;
    ctx.beginPath();
    ctx.moveTo(w / 2 + 10, 18);
    ctx.lineTo(w / 2 + 10, 54);
    ctx.stroke();
    ctx.restore();

    // Timer (right half)
    ctx.save();
    ctx.font         = 'bold 38px system-ui, sans-serif';
    ctx.fillStyle    = urgent ? '#FF5020' : '#FFE040';
    ctx.textAlign    = 'right';
    ctx.textBaseline = 'middle';
    ctx.shadowColor  = urgent ? 'rgba(255,80,20,0.5)' : 'transparent';
    ctx.shadowBlur   = urgent ? 10 : 0;
    ctx.fillText(timerText, panelX + panelW - 16, 37);
    ctx.restore();

    // ── Exit — top-right ───────────────────────────────────────────────────
    const [ex1, ey1, ex2, ey2] = this._btnActiveExit(w, h);
    drawBtn(ctx, ex1, ey1, ex2, ey2, '← Exit', 'rgba(210, 40, 40, 0.88)', { fontSize: 17 });
  }

  _drawEnd(ctx, w, h) {
    // Dark overlay
    ctx.save();
    ctx.globalAlpha = 0.6;
    ctx.fillStyle   = 'rgba(8, 8, 18, 1)';
    ctx.fillRect(0, 0, w, h);
    ctx.restore();

    const cx = w / 2, cy = h / 2;

    // Card panel
    drawPanel(ctx, cx - 220, cy - 130, 440, 280, 22);

    // Title
    ctx.save();
    ctx.font          = 'bold 58px system-ui, sans-serif';
    ctx.fillStyle     = '#FFE040';
    ctx.textAlign     = 'center';
    ctx.textBaseline  = 'alphabetic';
    ctx.shadowColor   = 'rgba(255,220,0,0.4)';
    ctx.shadowBlur    = 16;
    ctx.fillText('Game Over!', cx, cy - 48);
    ctx.restore();

    // Score
    ctx.save();
    ctx.font          = 'bold 40px system-ui, sans-serif';
    ctx.fillStyle     = '#fff';
    ctx.textAlign     = 'center';
    ctx.textBaseline  = 'alphabetic';
    ctx.fillText(`Score: ${this.score}`, cx, cy + 14);
    ctx.restore();

    // Restart button
    const [rx1, ry1, rx2, ry2] = this._btnRestart(w, h);
    drawBtn(ctx, rx1, ry1, rx2, ry2, 'Restart', 'rgba(30, 170, 60, 0.90)', { fontSize: 20 });

    // Exit button
    const [bx1, by1, bx2, by2] = this._btnExit(w, h);
    drawBtn(ctx, bx1, by1, bx2, by2, 'Exit', 'rgba(210, 40, 40, 0.90)', { fontSize: 20 });

    // Prompt
    ctx.save();
    ctx.font         = '15px system-ui, sans-serif';
    ctx.fillStyle    = 'rgba(160,160,180,0.9)';
    ctx.textAlign    = 'center';
    ctx.textBaseline = 'alphabetic';
    ctx.fillText('Point your index finger at a button', cx, cy + 120);
    ctx.restore();
  }
}
